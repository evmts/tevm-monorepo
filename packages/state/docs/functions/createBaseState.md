[**@tevm/state**](../README.md) • **Docs**

***

[@tevm/state](../globals.md) / createBaseState

# Function: createBaseState()

`Internal`

> **createBaseState**(`options`): [`BaseState`](../type-aliases/BaseState.md)

Creates the core data structure for ethereum state

## Parameters

• **options**: [`StateOptions`](../type-aliases/StateOptions.md)

## Returns

[`BaseState`](../type-aliases/BaseState.md)

Creates the core data structure the state manager operates on

## Source

[packages/state/src/createBaseState.js:20](https://github.com/evmts/tevm-monorepo/blob/main/packages/state/src/createBaseState.js#L20)
