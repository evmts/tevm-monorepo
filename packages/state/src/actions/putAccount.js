/**
 * Saves an account into state under the provided `address`.
 * @type {import("../state-types/index.js").StateAction<'putAccount'>}
 */
export const putAccount = (baseState) => async (address, account) => {
	if (account !== undefined) {
		baseState.caches.accounts?.put(address, account)
	} else {
		baseState.caches.accounts?.del(address)
	}
}
