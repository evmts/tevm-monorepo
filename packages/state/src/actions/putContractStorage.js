import { getAccount } from './getAccount.js'

/**
 * @param {Uint8Array} bytes
 * @returns {Uint8Array}
 */
const stripZeros = (bytes) => {
	if (!(bytes instanceof Uint8Array)) {
		throw new Error('Unexpected type')
	}
	return bytes.slice(bytes.findIndex(/** @param {number} entry*/ (entry) => entry !== 0))
}
/**
 * Adds value to the cache for the `account`
 * corresponding to `address` at the provided `key`.
 * Cannot be more than 32 bytes. Leading zeros are stripped.
 * If it is empty or filled with zeros, deletes the value.
 * @type {import("../state-types/index.js").StateAction<'putContractStorage'>}
 */
export const putContractStorage = (baseState) => async (address, key, value) => {
	if (key.length !== 32) {
		throw new Error('Storage key must be 32 bytes long!')
	}

	const account = await getAccount(baseState)(address)
	if (!account) {
		throw new Error('cannot putContractStorage on non existing acccount!')
	}
	baseState.caches.storage.put(address, key, stripZeros(value))
}
