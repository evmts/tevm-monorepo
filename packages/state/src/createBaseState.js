import { AccountCache, CacheType, StorageCache } from '@ethereumjs/statemanager'
import { createLogger } from '@tevm/logger'
import { ContractCache } from './ContractCache.js'
import { checkpoint } from './actions/checkpoint.js'
import { commit } from './actions/commit.js'
import { generateCanonicalGenesis } from './actions/generateCannonicalGenesis.js'

/**
 * @type {import('viem').Hex}
 */
const INITIAL_STATE_ROOT = '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421'

/**
 * @internal
 * Creates the core data structure for ethereum state
 * @param {import('./state-types/index.js').StateOptions} options
 * @returns {import('./BaseState.js').BaseState}
 * Creates the core data structure the state manager operates on
 */
export const createBaseState = (options) => {
	const logger = createLogger({
		level: options.loggingLevel ?? 'warn',
		name: '@tevm/state-manager',
	})
	/**
	 * @type {import('./state-types/StateRoots.js').StateRoots}
	 */
	const stateRoots = options.stateRoots ?? new Map()
	if (options.genesisState && options.currentStateRoot) {
		stateRoots.set(options.currentStateRoot, options.genesisState)
	} else {
		stateRoots.set(INITIAL_STATE_ROOT, options.genesisState ?? {})
	}
	let currentStateRoot = options.currentStateRoot ?? INITIAL_STATE_ROOT
	/**
	 * @type {import('./BaseState.js').BaseState}
	 */
	const state = {
		logger,
		getCurrentStateRoot: () => currentStateRoot,
		setCurrentStateRoot: (root) => {
			if (!state.stateRoots.has(root)) {
				throw new Error('Cannot set state root to non existing state root')
			}
			currentStateRoot = root
		},
		stateRoots,
		options,
		caches: {
			contracts: new ContractCache(),
			accounts: new AccountCache({
				size: 100_000,
				type: CacheType.ORDERED_MAP,
			}),
			storage: new StorageCache({
				size: 100_000,
				type: CacheType.ORDERED_MAP,
			}),
		},
		ready: () => genesisPromise.then(() => true),
	}
	const genesisPromise = (
		options.genesisState !== undefined && options.currentStateRoot === undefined
			? generateCanonicalGenesis(state)(options.genesisState)
			: Promise.resolve().then(() => {
					if (options.currentStateRoot) {
						state.setCurrentStateRoot(options.currentStateRoot)
						return generateCanonicalGenesis(state)(options.genesisState ?? stateRoots.get(options.currentStateRoot))
					}
					return Promise.resolve()
				})
	).then(async () => {
		await checkpoint(state)()
		await commit(state)()
		logger.debug('StateManager is ready')
	})

	return state
}
