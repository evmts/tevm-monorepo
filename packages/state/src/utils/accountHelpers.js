import { Rlp } from '@tevm/rlp'
import { createAccount, createAddressFromString as createAddress } from '@tevm/utils'

/**
 * Helper to create an address from a string
 * @param {string} addressString - The address string
 * @returns {import('@tevm/utils').EthjsAddress} The address
 */
export function createAddressFromString(addressString) {
	return createAddress(addressString)
}

/**
 * Creates an account from RLP serialized data
 * @param {Uint8Array} serialized - The RLP serialized account data
 * @returns {import('@tevm/utils').EthjsAccount} The account
 */
export function fromRlpSerializedAccount(serialized) {
	const decoded = Rlp.decode(serialized)
	if (!Array.isArray(decoded) || decoded.length !== 4) {
		throw new Error('Invalid RLP serialized account')
	}

	const [nonce, balance, storageRoot, codeHash] = decoded

	/** @type {{nonce: any, balance: any, storageRoot: any, codeHash: any}} */
	const accountData = {
		nonce,
		balance,
		storageRoot,
		codeHash,
	}

	return createAccount(accountData)
}

/**
 * Creates an account from account data
 * @param {{nonce?: bigint, balance?: bigint, storageRoot?: Uint8Array, codeHash?: Uint8Array}} accountData - The account data
 * @returns {import('@tevm/utils').EthjsAccount} The account
 */
export function fromAccountData(accountData) {
	return createAccount(accountData)
}
