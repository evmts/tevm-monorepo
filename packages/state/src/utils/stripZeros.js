import { InternalError } from '@tevm/errors'

/**
 * @param {Uint8Array} bytes
 * @returns {Uint8Array}
 */
export const stripZeros = (bytes) => {
	if (!(bytes instanceof Uint8Array)) {
		throw new InternalError('Unexpected type')
	}
	return bytes.slice(bytes.findIndex(/** @param {number} entry*/ (entry) => entry !== 0))
}
