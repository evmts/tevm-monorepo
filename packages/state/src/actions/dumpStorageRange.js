import { InvalidParamsError } from '@tevm/errors'
import { bytesToHex, getAddress, hexToBigInt } from '@tevm/utils'

/**
 * @type {import("../state-types/index.js").StateAction<'dumpStorageRange'>}
 */
export const dumpStorageRange = (state) => (_address, _startKey, _limit) => {
	const storage = state.caches.storage.dump(_address)
	if (!storage) {
		throw new InvalidParamsError(`No storage found at address ${getAddress(_address.toString())}`)
	}
	/**
	 * @type {Array<[import("viem").Hex, {key: import("@tevm/utils").Hex, value: import('@tevm/utils').Hex}]>}
	 */
	const entries = []
	/**
	 * @type {import("viem").Hex | null}
	 */
	let nextKey = null
	let started = false
	for (const [storageKey, storageValue] of storage.entries()) {
		if (entries.length === _limit) {
			nextKey = /** @type {import("@tevm/utils").Hex}*/ (storageKey)
			break
		}
		if (!started) {
			if (hexToBigInt(/** @type {import('@tevm/utils').Hex}*/ (storageKey)) === _startKey) {
				started = true
			} else {
				continue
			}
		}
		const key = /** @type {import("@tevm/utils").Hex}*/ (storageKey)
		const value = bytesToHex(storageValue)
		entries.push([key, { key, value }])
	}
	return Promise.resolve({
		nextKey,
		storage: Object.fromEntries(entries),
	})
}
