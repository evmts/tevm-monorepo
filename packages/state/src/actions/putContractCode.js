import { EthjsAccount, keccak256 } from '@tevm/utils'
import { getAccount } from './getAccount.js'
import { putAccount } from './putAccount.js'

/**
 * Adds `value` to the state trie as code, and sets `codeHash` on the account
 * corresponding to `address` to reference this.
 * @type {import("../state-types/index.js").StateAction<'putContractCode'>}
 */
export const putContractCode = (baseState) => async (address, value) => {
	const account = await getAccount(baseState)(address)
	await putAccount(baseState)(
		address,
		EthjsAccount.fromAccountData({
			...account,
			codeHash: keccak256(value, 'bytes'),
		}),
	)
	baseState.caches.contracts.put(address, value)
	return
}
