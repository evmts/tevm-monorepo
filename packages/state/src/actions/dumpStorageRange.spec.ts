import { createAddress } from '@tevm/address'
import { SimpleContract } from '@tevm/test-utils'
import { EthjsAccount } from '@tevm/utils'
import { type Hex, hexToBigInt, hexToBytes, numberToBytes } from 'viem'
import { describe, expect, it } from 'vitest'
import { createBaseState } from '../createBaseState.js'
import { dumpStorageRange } from './dumpStorageRange.js'
import { putAccount } from './putAccount.js'
import { putContractCode } from './putContractCode.js'
import { putContractStorage } from './putContractStorage.js'

describe(dumpStorageRange.name, () => {
	it('should dump storage from a range', async () => {
		const state = createBaseState({})
		const address = createAddress(9876543210)
		await putAccount(state)(
			address,
			EthjsAccount.fromAccountData({
				nonce: 2n,
				balance: 420n,
			}),
		)
		await putContractCode(state)(address, hexToBytes(SimpleContract.deployedBytecode))
		await putContractStorage(state)(address, numberToBytes(0, { size: 32 }), numberToBytes(420, { size: 32 }))
		expect(await dumpStorageRange(state)(address, 0n, 20)).toMatchSnapshot()
	})

	it('supports start key and limit', async () => {
		const state = createBaseState({})
		const address = createAddress(9876543210)
		await putAccount(state)(
			address,
			EthjsAccount.fromAccountData({
				nonce: 2n,
				balance: 420n,
			}),
		)
		await putContractCode(state)(address, hexToBytes(SimpleContract.deployedBytecode))
		await putContractStorage(state)(address, numberToBytes(0, { size: 32 }), numberToBytes(420, { size: 32 }))
		await putContractStorage(state)(address, numberToBytes(1, { size: 32 }), numberToBytes(420, { size: 32 }))
		await putContractStorage(state)(address, numberToBytes(2, { size: 32 }), numberToBytes(420, { size: 32 }))
		await putContractStorage(state)(address, numberToBytes(3, { size: 32 }), numberToBytes(420, { size: 32 }))
		await putContractStorage(state)(address, numberToBytes(4, { size: 32 }), numberToBytes(420, { size: 32 }))
		await putContractStorage(state)(address, numberToBytes(5, { size: 32 }), numberToBytes(420, { size: 32 }))
		await putContractStorage(state)(address, numberToBytes(6, { size: 32 }), numberToBytes(420, { size: 32 }))

		const getPaginatedStorage = async () => {
			let startKey: bigint | null = 0n
			const limit = 2
			const storage = {}
			while (startKey !== null) {
				const nextStorage = await dumpStorageRange(state)(address, startKey, limit)
				startKey = nextStorage.nextKey === null ? null : hexToBigInt(nextStorage.nextKey as Hex)
				Object.assign(storage, nextStorage.storage)
			}
			return storage
		}

		expect(await dumpStorageRange(state)(address, 0n, 20)).toMatchInlineSnapshot(`
			{
			  "nextKey": null,
			  "storage": {
			    "0000000000000000000000000000000000000000000000000000000000000000": {
			      "key": "0000000000000000000000000000000000000000000000000000000000000000",
			      "value": "0x01a4",
			    },
			    "0000000000000000000000000000000000000000000000000000000000000001": {
			      "key": "0000000000000000000000000000000000000000000000000000000000000001",
			      "value": "0x01a4",
			    },
			    "0000000000000000000000000000000000000000000000000000000000000002": {
			      "key": "0000000000000000000000000000000000000000000000000000000000000002",
			      "value": "0x01a4",
			    },
			    "0000000000000000000000000000000000000000000000000000000000000003": {
			      "key": "0000000000000000000000000000000000000000000000000000000000000003",
			      "value": "0x01a4",
			    },
			    "0000000000000000000000000000000000000000000000000000000000000004": {
			      "key": "0000000000000000000000000000000000000000000000000000000000000004",
			      "value": "0x01a4",
			    },
			    "0000000000000000000000000000000000000000000000000000000000000005": {
			      "key": "0000000000000000000000000000000000000000000000000000000000000005",
			      "value": "0x01a4",
			    },
			    "0000000000000000000000000000000000000000000000000000000000000006": {
			      "key": "0000000000000000000000000000000000000000000000000000000000000006",
			      "value": "0x01a4",
			    },
			  },
			}
		`)
		expect(await getPaginatedStorage()).toMatchInlineSnapshot(`
			{
			  "0000000000000000000000000000000000000000000000000000000000000000": {
			    "key": "0000000000000000000000000000000000000000000000000000000000000000",
			    "value": "0x01a4",
			  },
			  "0000000000000000000000000000000000000000000000000000000000000001": {
			    "key": "0000000000000000000000000000000000000000000000000000000000000001",
			    "value": "0x01a4",
			  },
			  "0000000000000000000000000000000000000000000000000000000000000002": {
			    "key": "0000000000000000000000000000000000000000000000000000000000000002",
			    "value": "0x01a4",
			  },
			  "0000000000000000000000000000000000000000000000000000000000000003": {
			    "key": "0000000000000000000000000000000000000000000000000000000000000003",
			    "value": "0x01a4",
			  },
			  "0000000000000000000000000000000000000000000000000000000000000004": {
			    "key": "0000000000000000000000000000000000000000000000000000000000000004",
			    "value": "0x01a4",
			  },
			  "0000000000000000000000000000000000000000000000000000000000000005": {
			    "key": "0000000000000000000000000000000000000000000000000000000000000005",
			    "value": "0x01a4",
			  },
			  "0000000000000000000000000000000000000000000000000000000000000006": {
			    "key": "0000000000000000000000000000000000000000000000000000000000000006",
			    "value": "0x01a4",
			  },
			}
		`)
	})
})
