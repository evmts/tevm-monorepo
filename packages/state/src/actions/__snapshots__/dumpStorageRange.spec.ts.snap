// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dumpStorageRange > should dump storage from a range 1`] = `
{
  "nextKey": null,
  "storage": {
    "0000000000000000000000000000000000000000000000000000000000000000": {
      "key": "0000000000000000000000000000000000000000000000000000000000000000",
      "value": "0x01a4",
    },
  },
}
`;
